### ISSUE TITLE
Use NetworkExtension API to on macOS rather than direct tun interface

### ISSUE THREAD

**Issue by liilac** [oxen-io/lokinet#117]
NOTE: This is being worked on by me already as part of my iOS work.

At present, root access is required to run lokinet on Darwin platforms including macOS.  There are unprivileged APIs which can be used to avoid this, which further provide deep platform integration and other benefits.

See [NetworkExtension API](https://developer.apple.com/documentation/networkextension) docs for reference.

There are other ways to do this without requiring root, but IMO NE API is by far the best choice.  It's also required for iOS support (see #62), so I'll be implementing it regardless.

Not sure whether plain Darwin (i.e. not macOS) includes the NetworkExtension APIs at present, will have to check.  Is this a platform we plan to support?

**Comment by majestrate** [comment]
one associated worry i have is that this will add a hard requirement on having code signed binaries on apple platforms, which makes compiling it harder.

**Comment by liilac** [comment]
I think that fear is a bit misplaced.

The requirement always exists for iOS, no matter what, but code signing certificates are readily available.  The project can even distribute outside the app store with enterprise distribution channels (the iOS app can be a build artefact on a repo, even, and users can install from their browser).  This requires those compiling to have access to a code signing certificate, however, which can cost $99 per year for some.

For macOS, using the NetworkExtension API means in a system configured following security best-practices, code signing will be required, however, there is nothing stopping users from singing with a **self-signed certificate** they have installed to their own machine.

TL;DR self-signed certs can be used for macOS, this is a non-issue in practice

@majestrate

**Comment by majestrate** [comment]
fixed in recent dev branch. just need to wire it up to a gui.

**Comment by barracuda156** [comment]
@liilac No, Darwin won’t have this, and requiring NE in effect has broken all macOS support prior to 10.10 if not 10.15.

I am trying to get rid of NE now.
